### ISSUE TITLE
F821 false positive for using a builtin after deleting a name that shadowed the builtin

### ISSUE THREAD

**Issue by dscorbett** [astral-sh/ruff#14772]
[`undefined-name` (F821)](https://docs.astral.sh/ruff/rules/undefined-name/) in Ruff 0.8.1 reports a name as undefined if it is used after being deleted but the name was shadowing a builtin. In that case, the name will resolve to the builtin without a `NameError`, so the rule should not report a problem.

```console
$ cat f821_1.py
print = 0
del print
print(1)

$ python f821_1.py
1

$ ruff check --isolated --select F821 f821_1.py
f821_1.py:3:1: F821 Undefined name `print`. Consider specifying `requires-python = ">= 3.0"` or `tool.ruff.target-version = "py30"` in your `pyproject.toml` file.
  |
1 | print = 0
2 | del print
3 | print(1)
  | ^^^^^ F821
  |

Found 1 error.
```

Specifying the target version as suggested gives a different error.

```console
$ ruff check --isolated --select F821 f821_1.py --config 'target-version = "py30"'
error: invalid value 'target-version = "py30"' for '--config <CONFIG_OPTION>'

  tip: A `--config` flag must either be a path to a `.toml` configuration file
       or a TOML `<KEY> = <VALUE>` pair overriding a specific configuration
       option

Could not parse the supplied argument as a `ruff.toml` configuration option:

unknown variant `py30`, expected one of `py37`, `py38`, `py39`, `py310`, `py311`, `py312`, `py313`
in `target-version`

For more information, try '--help'.
```

That’s just a minimal example. Here is more realistic example based on [the documentation of the `gettext` module](https://docs.python.org/3/library/gettext.html#deferred-translations), which recommends this pattern for deferred translations.

```console
$ cat f821_2.py
import gettext
gettext.install("example")
def _(message):
    return message
messages = [_("x"), _("y"), _("z")]
del _
for message in messages:
    print(_(message))

$ python f821_2.py
x
y
z

$ ruff check --isolated --select F821 f821_2.py
f821_2.py:8:11: F821 Undefined name `_`
  |
6 | del _
7 | for message in messages:
8 |     print(_(message))
  |           ^ F821
  |

Found 1 error.
```

That example fails even when `_` is configured as a builtin.

```console
$ ruff check --isolated --select F821 f821_2.py --config 'builtins = ["_"]'
f821_2.py:8:11: F821 Undefined name `_`
  |
6 | del _
7 | for message in messages:
8 |     print(_(message))
  |           ^ F821
  |

Found 1 error.
```

**Comment by dylwil3** [comment]
That's an interesting and subtle find! I didn't know about this special treatment of builtins. A related thing happens if you try to delete a built-in:

```console
❯ uv run --no-project python
Python 3.12.7 (main, Oct 16 2024, 07:12:08) [Clang 18.1.8 ] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> print
<built-in function print>
>>> del print
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'print' is not defined
```

Sort of curious...

Neither [pyright](https://pyright-play.net/?pythonVersion=3.12&code=A4JwlgdgLgBAvDAjAKACYFMA2NSSsgD3h3GiA) nor [mypy](https://mypy-play.net/?mypy=latest&python=3.12&gist=892483f184ba77b480e95bd8fe9ba21b) seem to know about these facts either.

Does `red-knot` know about this @carljm ?

**Comment by carljm** [comment]
Yeah, mypy and pyright both look pretty confused here :) `del` isn't very common, and even less common with a name that is also a builtin, so it probably doesn't come up often.

> Does `red-knot` know about this [@carljm](https://github.com/carljm) ?

Red-knot doesn't understand `del` at all yet, so no :) But if we implement `del` support in the way I expect we would, I think correct handling of this should fall out naturally from our model.

**Comment by dylwil3** [comment]
> so it probably doesn't come up often.
Yeah I don't think this should be a high priority to support 😄 though it sounds like you're saying it may happen automatically

> FWIW, in terms of what's happening at runtime, this isn't really "special treatment of builtins". The del statement is working on a name that shadows a builtin in exactly the same way it works on any other name.

I think what I found surprising/special is the console example I had above:

```console
❯ uv run --no-project python
Python 3.12.7 (main, Oct 16 2024, 07:12:08) [Clang 18.1.8 ] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> print
<built-in function print>
>>> del print
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'print' is not defined
```

Specifically, it's strange to me that the error it's giving me is that `'print' is not defined`. But maybe what you're saying is that NameError's form CPython have an implicit "within the module scope" in them.

Anyway this is pretty obscure, so no need to keep letting me distract you with it 😄

**Comment by carljm** [comment]
> maybe what you're saying is that NameError's form CPython have an implicit "within the module scope" in them.

In general, no: typically for a name _reference_ you'd only get a `NameError` if we can't find the name, even after looking up in enclosing scopes / builtins (if not defined in local scope.)

But for the `NameError` in `del`, yes, that's exactly right.

So I think really what I'm saying is "it's `del` that's weird, not builtins" :)
